src/lib/core/datetime/date-adapter.ts

```diff
@@ -192,6 +185,23 @@ export abstract class DateAdapter<D> {
   abstract isValid(date: D): boolean;
 
   /**
+   * Attempts to coerce a value to a valid date object. This is different from parsing in that it
+   * should only coerce non-ambiguous, locale-independent values (e.g. a ISO 8601 string).
+   * The default implementation does not allow any coercion, it simply checks that the given value
+   * is already a valid date object or null.
+   * @param value The value to be coerced to a date object.
+   * @returns The coerced date object, either a valid date, null if the value can be coerced to a
+   *     null date (e.g. the empty string).
+   * @throws If the given value cannot be coerced to a valid date or null.
+   */
+  coerceToDate(value: any): D | null {
```

Comment:
Doesn't this need to align with `toIso8601`? I think the heart of what this is trying to do is the same as `parse` / `format` except that instead of being string that the user see / inputs, it's the formats that the backend sends / accepts. Looking at it that way, I think this would make sense:

`parse` / `format` -> for **user-facing** strings
`serialize` / `deserialize` -> for **backend-facing** strings (or JSON or whatever)

Reply:
For user-facing strings: I agree we need both.

For backend-facing strings: I don't think the `DateAdapter` is particularly concerned with these for the most part. We don't really know if the dates coming into / out of the datepicker are going to the backend or some other part of the JS app, its up to the user to decide and handle that. However for convenience we might want to allow certain values (such as ISO 8601 strings or UNIX timestamps) to be automatically coerced into the proper type - similar to how we coerce `"true"` and `"false"` for boolean properties. This is what the new `coerceToDate` method is about, allowing the user to decide what can be coerced and what can't.

There's also a third class: DOM-facing strings. For example to set the `min` attribute on the input we need the date in a format that the DOM can understand, which is ISO 8601. This is why we always need to have a way to go from whatever the date type is to an ISO 8601 string.

Reply:
How do you feel, then, about calling this method `deserialize`? 

Also on `toIso8601`, we should document somewhere why/how this is used.

Reply:
👍 

I'll update the description of all the to/from date methods to make sure its clear exactly what each one is used for